keps/sig-node/2238-liveness-probe-grace-period/README.md

```diff
@@ -329,17 +322,34 @@ _This section must be completed when targeting beta graduation to a release._
   Try to be as paranoid as possible - e.g., what if some components will restart
    mid-rollout?
 
+  This change is purely additive. Already-running workloads on a node will not
+  have this field set.
+
 * **What specific metrics should inform a rollback?**
 
+  Increases in error rates on shutdowns initiated by probes; unexpected
+  shutdown times for pods with containers that have probes configured.
+
 * **Were upgrade and rollback tested? Was the upgrade->downgrade->upgrade path tested?**
   Describe manual testing that was done and the outcomes.
   Longer term, we may want to require automated upgrade/rollback tests, but we
   are missing a bunch of machinery and tooling and can't do that now.
 
+  For the API server, rollback can be tested by:
+
+  - Turning on the feature flag
+  - Creating a workload with the field
+  - Disabling the feature flag
+
+  We will remove the feature flag from the kubelet for beta to avoid worrying
```

Comment:
If the field is set, and then the flag disabled in apiserver, I assume the field will not be present when the kubelet reads the spec?

Reply:
Right now, it's only unset when the feature flag is disabled *and* the field is not present, looks like: https://github.com/kubernetes/kubernetes/blob/6296c6e4d336f9d3e4cc80e0af5d4bc74f330d0d/pkg/api/pod/util.go#L553-L565

Would I need to make this change?

Reply:
Well, if we don't, then wouldn't we still need the feature flag in kubelet? Turning off the feature at the api server alone won't disable it if the kubelet still gets the value. What's the standard practice for this in the API?

Reply:
I'm not sure. @derekwaynecarr wdyt?

Reply:
I think it is safe to remove the feature gate from the kubelet under beta criteria.

This starts the clock for all nodes to understand this new field.
In two releases, we default the kube-apiserver to enable the feature gate so all environments will have kubelets connected that understand the feature natively.

For users that want to exercise the new field earlier, they need to enable the feature gate on the kube-apiserver and ensure all connected kubelets are at equal level N.

Reply:
See my question below wrt enable/disable scenario.

Reply:
> Turning off the feature at the api server alone won't disable it if the kubelet still gets the value. What's the standard practice for this in the API?

Turning off a feature flag does not clear previously persisted data from the API field, it prevents data in the field in new objects (or in updates of existing objects that don't already use the field).

Reply:
Right, so my point is that we can't remove the kubelet feature flag, because if we do, then the user has no control over those old objects still containing the value, and no way to then back out the feature on those objects already with the value (in the enable then disable scenario). Except to delete it entirely and recreated it. 

So, the question is: if we want to remove the kubelet feature flag early, would it be OK to blank the field on output from the api server when the feature flag is disabled? That would prevent the kubelet from seeing the new value and effectively disable the functionality without having to keep the kubelet feature flag.

Reply:
@johnbelamaric that makes sense.  the other option is if we could say the feature flag defaulted to on for kubelet, but off for kube-apiserver, but I am not aware of when we have done that in the past (or if we did, my memory escapes me).

Reply:
I think the easiest way forward is probably to blank out the field when the feature flag is off. I will update.

Reply:
Ok, let's move forward with this as plan-of-record, if Jordan or other API reviewers hate it we can revisit...

Reply:
this may be possible, i just cant remember what i did on this and if it handled the write side but not the read side.